File: packages/core/src/routes/datasets.ts
import { Hono } from 'hono';
import { getDb, schema } from '../lib/db.js';
import { eq, desc } from 'drizzle-orm';
import fs from 'node:fs/promises';
import path from 'node:path';
import { env } from '../lib/env.js';

export const datasetsRouter = new Hono();
const userIdFromCtx = () => 1;

// List datasets
datasetsRouter.get('/', async (c) => {
  const db = await getDb();
  const id = userIdFromCtx();
  const rows = await db.select().from(schema.datasets)
    .where(eq(schema.datasets.user_id, id))
    .orderBy(desc(schema.datasets.created_at));
  return c.json(rows);
});

// Get dataset details
datasetsRouter.get('/:id', async (c) => {
  const db = await getDb();
  const dsId = parseInt(c.req.param('id'));
  const [ds] = await db.select().from(schema.datasets).where(eq(schema.datasets.id, dsId)).limit(1);
  if (!ds) return c.json({ error: 'Not found' }, 404);
  return c.json(ds);
});

// Create dataset from CSV/JSON upload
datasetsRouter.post('/', async (c) => {
  const db = await getDb();
  const id = userIdFromCtx();
  const body = await c.req.json();
  const { name, description, data, filename } = body;

  if (!name || !data) return c.json({ error: 'name and data required' }, 400);

  const dsDir = path.join(env.WORKSPACE_DIR, 'datasets');
  await fs.mkdir(dsDir, { recursive: true });
  const slug = name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/(^-|-$)/g, '');
  const dbPath = path.join(dsDir, `${slug}.duckdb`);

  try {
    // Write the raw data to a temp file for DuckDB to ingest
    const ext = (filename || 'data.csv').split('.').pop()?.toLowerCase() || 'csv';
    const tmpFile = path.join(dsDir, `_tmp_${slug}.${ext}`);
    await fs.writeFile(tmpFile, data, 'utf8');

    // Use DuckDB to ingest data
    const duckdb = await import('duckdb-async');
    const ddb = await duckdb.Database.create(dbPath);
    const conn = await ddb.connect();

    let tableName = 'data';
    if (ext === 'json') {
      await conn.run(`CREATE TABLE "${tableName}" AS SELECT * FROM read_json_auto('${tmpFile}')`);
    } else {
      await conn.run(`CREATE TABLE "${tableName}" AS SELECT * FROM read_csv_auto('${tmpFile}')`);
    }

    // Get schema and row count
    const schemaRows = await conn.all(`SELECT column_name, data_type FROM information_schema.columns WHERE table_name='${tableName}'`);
    const countRows = await conn.all(`SELECT count(*) as cnt FROM "${tableName}"`);
    const rowCount = Number(countRows[0]?.cnt || 0);

    await conn.close();
    await ddb.close();

    // Clean up temp file
    await fs.unlink(tmpFile).catch(() => {});

    const schemaDef = schemaRows.map((r: any) => ({ name: r.column_name, type: r.data_type }));

    const [ds] = await db.insert(schema.datasets).values({
      user_id: id,
      name,
      description: description || '',
      source: filename || 'upload',
      schema_def: schemaDef,
      row_count: rowCount,
      file_path: dbPath,
    } as any).returning();

    return c.json(ds);
  } catch (err: any) {
    return c.json({ error: err.message }, 500);
  }
});

// Delete dataset
datasetsRouter.delete('/:id', async (c) => {
  const db = await getDb();
  const dsId = parseInt(c.req.param('id'));
  const [ds] = await db.select().from(schema.datasets).where(eq(schema.datasets.id, dsId)).limit(1);
  if (!ds) return c.json({ error: 'Not found' }, 404);

  // Delete DuckDB file
  if (ds.file_path) {
    await fs.unlink(ds.file_path).catch(() => {});
  }

  await db.delete(schema.datasets).where(eq(schema.datasets.id, dsId));
  return c.json({ ok: true });
});

// Run SQL query against dataset
datasetsRouter.post('/:id/query', async (c) => {
  const db = await getDb();
  const dsId = parseInt(c.req.param('id'));
  const [ds] = await db.select().from(schema.datasets).where(eq(schema.datasets.id, dsId)).limit(1);
  if (!ds) return c.json({ error: 'Not found' }, 404);

  const body = await c.req.json();
  const { sql } = body;
  if (!sql) return c.json({ error: 'sql required' }, 400);

  try {
    const duckdb = await import('duckdb-async');
    const ddb = await duckdb.Database.create(ds.file_path!);
    const conn = await ddb.connect();
    const rows = await conn.all(sql);
    await conn.close();
    await ddb.close();
    return c.json({ rows, columns: rows.length > 0 ? Object.keys(rows[0]) : [] });
  } catch (err: any) {
    return c.json({ error: err.message }, 400);
  }
});

// Preview first 100 rows
datasetsRouter.get('/:id/preview', async (c) => {
  const db = await getDb();
  const dsId = parseInt(c.req.param('id'));
  const [ds] = await db.select().from(schema.datasets).where(eq(schema.datasets.id, dsId)).limit(1);
  if (!ds) return c.json({ error: 'Not found' }, 404);

  try {
    const duckdb = await import('duckdb-async');
    const ddb = await duckdb.Database.create(ds.file_path!);
    const conn = await ddb.connect();
    const rows = await conn.all('SELECT * FROM data LIMIT 100');
    await conn.close();
    await ddb.close();
    return c.json({ rows, columns: rows.length > 0 ? Object.keys(rows[0]) : [] });
  } catch (err: any) {
    return c.json({ error: err.message }, 500);
  }
});

File: packages/api/app/routers/datasets.py
from pathlib import Path

import duckdb
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.dataset import Dataset
from app.models.user import User
from app.services import dataset_manager

router = APIRouter(prefix="/api/datasets", tags=["datasets"])


class DatasetCreate(BaseModel):
    name: str
    description: str = ""
    data: str
    filename: str = "data.csv"


class QueryRequest(BaseModel):
    sql: str


@router.get("/")
async def list_datasets(user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    return await dataset_manager.list_datasets(db, user.id)


@router.get("/{dataset_id}")
async def get_dataset(dataset_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    ds = await dataset_manager.get_dataset(db, user.id, dataset_id)
    if not ds:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return ds


@router.post("/")
async def create_dataset(body: DatasetCreate, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    ds_dir = Path(settings.workspace_dir) / "datasets"
    ds_dir.mkdir(parents=True, exist_ok=True)
    slug = "".join(c if c.isalnum() else "-" for c in body.name.lower()).strip("-")
    db_path = ds_dir / f"{slug}.duckdb"

    ext = body.filename.rsplit(".", 1)[-1].lower() if "." in body.filename else "csv"
    tmp_file = ds_dir / f"_tmp_{slug}.{ext}"
    tmp_file.write_text(body.data)

    try:
        con = duckdb.connect(str(db_path))
        if ext == "json":
            con.execute(f"CREATE TABLE data AS SELECT * FROM read_json_auto('{tmp_file}')")
        else:
            con.execute(f"CREATE TABLE data AS SELECT * FROM read_csv_auto('{tmp_file}')")

        schema_rows = con.execute(
            "SELECT column_name, data_type FROM information_schema.columns WHERE table_name='data'"
        ).fetchall()
        count_row = con.execute("SELECT count(*) FROM data").fetchone()
        row_count = count_row[0] if count_row else 0
        con.close()
    finally:
        tmp_file.unlink(missing_ok=True)

    schema_def = [{"name": r[0], "type": r[1]} for r in schema_rows]
    ds = Dataset(
        user_id=user.id, name=body.name, description=body.description,
        source=body.filename, schema_def=schema_def, row_count=row_count,
        file_path=str(db_path),
    )
    db.add(ds)
    await db.commit()
    await db.refresh(ds)
    return ds


@router.delete("/{dataset_id}")
async def delete_dataset(dataset_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    deleted = await dataset_manager.delete_dataset(db, user.id, dataset_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Dataset not found")
    return {"ok": True}


@router.post("/{dataset_id}/query")
async def run_query(dataset_id: int, body: QueryRequest, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    ds = await dataset_manager.get_dataset(db, user.id, dataset_id)
    if not ds:
        raise HTTPException(status_code=404, detail="Dataset not found")
    try:
        con = duckdb.connect(str(ds.file_path), read_only=True)
        rows = con.execute(body.sql).fetchdf().to_dict(orient="records")
        columns = list(rows[0].keys()) if rows else []
        con.close()
        return {"rows": rows, "columns": columns}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{dataset_id}/preview")
async def preview_dataset(dataset_id: int, user: User = Depends(get_current_user), db: AsyncSession = Depends(get_db)):
    ds = await dataset_manager.get_dataset(db, user.id, dataset_id)
    if not ds:
        raise HTTPException(status_code=404, detail="Dataset not found")
    try:
        con = duckdb.connect(str(ds.file_path), read_only=True)
        rows = con.execute("SELECT * FROM data LIMIT 100").fetchdf().to_dict(orient="records")
        columns = list(rows[0].keys()) if rows else []
        con.close()
        return {"rows": rows, "columns": columns}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
